File: doc/codes/python/examples/api/command/lowlevel/bank.py
from autd3.commands import ChangePatternBank, ConfigPattern, PatternCompression, WritePatternBuffer, WritePatternCompressed
from autd3.geometry import Autd3, Geometry
from autd3.value import Intensity, LoopBehavior, PatternBank, SamplingConfig, TransitionMode

geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

bank = PatternBank.B0
index = 0
phases = geometry.phase_buffer()
intensities = Intensity.MAX
# ANCHOR: write
WritePatternBuffer(
    bank=bank,
    index=index,
    phases=phases,
    intensities=intensities,
)
# ANCHOR_END: write
config = SamplingConfig.FREQ_4K
size = 1
loop_behavior = LoopBehavior.Infinite
# ANCHOR: config
ConfigPattern(
    bank=bank,
    config=config,
    size=size,
    loop_behavior=loop_behavior,
)
# ANCHOR_END: config
transition_mode = TransitionMode.Immediate
# ANCHOR: change
ChangePatternBank(
    bank=bank,
    transition_mode=transition_mode,
)
# ANCHOR_END: change

p0 = geometry.phase_buffer()
p1 = geometry.phase_buffer()
p2 = geometry.phase_buffer()
p3 = geometry.phase_buffer()
patterns = [p0, p1, p2, p3]
index = 0
format = PatternCompression.PhaseHalf
intensity = Intensity.MAX
# ANCHOR: compressed
WritePatternCompressed(
    bank=bank,
    index=index,
    format=format,
    intensity=intensity,
    patterns=patterns,
)
# ANCHOR_END: compressed

File: doc/codes/rust/examples/api/command/lowlevel/bank.rs
use core::num::NonZeroU16;

use anyhow::Result;

use autd3_rs::commands::{
    ChangePatternBank, ConfigPattern, PatternCompression, WritePatternBuffer,
    WritePatternCompressed,
};
use autd3_rs::geometry::{Autd3, Geometry};
use autd3_rs::value::{Intensity, LoopBehavior, PatternBank, SamplingConfig, TransitionMode};
use autd3_rs_link_nop::Nop;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let bank = PatternBank::B0;
    let index = 0;
    let _phases = geometry.phase_buffer();
    let phases = &_phases;
    let intensities = Intensity::MAX;
    // ANCHOR: write
    WritePatternBuffer::new(bank, index, phases, intensities);
    // ANCHOR_END: write
    let config = SamplingConfig::FREQ_4K;
    let size = 1;
    let loop_behavior = LoopBehavior::Infinite;
    // ANCHOR: config
    ConfigPattern {
        bank,
        config,
        size,
        loop_behavior,
    };
    // ANCHOR_END: config
    let transition_mode = TransitionMode::Immediate;
    // ANCHOR: change
    ChangePatternBank {
        bank,
        transition_mode,
    };
    // ANCHOR_END: change

    let p0 = geometry.phase_buffer();
    let p1 = geometry.phase_buffer();
    let p2 = geometry.phase_buffer();
    let p3 = geometry.phase_buffer();
    let patterns = [Some(&p0[..]), Some(&p1[..]), Some(&p2[..]), Some(&p3[..])];
    let index = 0;
    let format = PatternCompression::PhaseHalf;
    // ANCHOR: compressed
    WritePatternCompressed {
        bank,
        index,
        format,
        patterns,
    };
    // ANCHOR_END: compressed
    Ok(())
}
